lib/events/complete.go

```diff
@@ -416,3 +452,38 @@ loop:
 	}
 	return nil
 }
+
+func (u *UploadCompleter) createIncompleteSession(sessionID session.ID) error {
+	path := filepath.Join(u.cfg.UnconfirmedSessionEndDir, sessionID.String())
```

Comment:
I see, so we're creating an empty file on disk named after the session ID as in indicator to go back and retry the session.

But then when we retry we have to stream the entire session again to compute what the end event should look like.

Why don't we just write the JSON-encoded end event to this file, so subsequent retries can emit the end event directly rather than having to compute it again?

Reply:
Because we don't have the event every time.
If the audit log backend is suffering a temporary failure, streaming can fail and we can't even validate if the session already has the `session.end` event. 

Saying that, we need to retry later to:
1. pull all session events and check whether we have a `session.end` or not
2. if audit log failed, create the file and retry later.
3. if session doesn't have the `session.end` event, try creating it. If create audit event operation fails we retry later. For this step we can implement the optimization you mentioned but we still need to deal with step 2.